### PR TITLE
Added support for the TMS320C64x architecture.

### PR DESCRIPTION
**Pull request by fotisl**

### CODE REVIEW

**Comment by aquynh** [comment]
@emoon, please review the commit for M68k. if you approve, i will cherry pick it

**Comment by aquynh** [comment]
a quick glance, and this looks quite clean, nice works! 

few things:

- please add your name to CREDITS.TXT
- what is the status of this arch? is it complete, or there are still instructions to be added in the future?
- i dont see any modification to bindings/const_generator.py ?
- please add this a new test for this arch to tests/test_basic.c & bindings/python/test_basic.py

thanks

**Comment by emoon** [comment]
M68K changes looks ok! :+1:

**Comment by fotisl** [comment]
Thanks!

- I have added my name to CREDITS.TXT
- All instructions have been implemented for the TMS320C64x architecture. There are some additional instructions for the TMS320C64x+ architecture, but I do not plan to implement them for the moment. Moreover, there are some instructions that are disassembled incorrectly by the TI disassembler and correctly by capstone.
- Forgot to add the diffs, fixed.
- Added both tests.

There is a major difference with the TI assembly. TI encodes the parallel information at the opcode of the previous instruction. So, if you have:
   ADD.D1 A0, A1, A2
|| AND.L2 B0, B1, B2
the information about the second instruction being executed in parallel with the first is not encoded in the AND opcode, but in the ADD opcode. Since this cannot be represented correctly when disassembling single instructions, the above will be disassembled as:
   ADD.D1 A0, A1, A2 ||
   AND.L2 B0, B1, B2
The '||' has been moved to the end of the previous instruction.

**Comment by aquynh** [comment]
what do you think about having this arch as TMS320, and C64x as a mode?

**Comment by fotisl** [comment]
You can use TMS320C6000 as the arch and C64x as a mode. AFAIK, other architectures are completely different. Should I make the changes or can you do it?

**Comment by aquynh** [comment]
no, just my question. let see if others have anything to comment on this.

**Comment by radare** [comment]
maybe it will be good to port radare2's c55x disassembler to capstone, and merge it in here

**Comment by fotisl** [comment]
@radare c55x belongs to the c5000 series. I think it's better to create a separate architecture since they are incompatible (it's the same as merging armv7 and armv8).

**Comment by radare** [comment]
Makes sense

**Comment by aquynh** [comment]
this looks very nice. i merged this into a new branch "tms320c64x", and will merge it back to the "next" branch after a bit more review. thanks.

**Comment by radare** [comment]
will next be ever released?

its been really a looong wait, and im not sure if we should include more archs when we are in “stabilizing mode” for more than 3 months

**Comment by radare** [comment]
Please rebase

**Comment by radare** [comment]
the disassembly should be in lowercase by default, external tools can make this uppercase later, like its done for the rest of archs

**Comment by radare** [comment]
im fuzzing your branch right now and i'll submit tests soon

**Comment by radare** [comment]
```
  0%  6069  000017b5          LDNW.D1T1 *A0++[0==12976== Conditional jump or move depends on uninitialised value(s)
==12976==    at 0x100017C2C: decodeToMCInst_4 (in /Users/pancake/prg/cs-tms320/tests/libcapstone.4.dylib)
==12976==    by 0x10001660A: TMS320C64x_getInstruction (in /Users/pancake/prg/cs-tms320/tests/libcapstone.4.dylib)
==12976==    by 0x10001391D: cs_disasm (in /Users/pancake/prg/cs-tms320/tests/libcapstone.4.dylib)
==12976==    by 0x100000E0F: test (in ./test_tms320c64x)
==12976==    by 0x100000CF3: main (in ./test_tms320c64x)
==12976==
], A0   ||
```

**Comment by radare** [comment]
Another bug:
```
==13592== Invalid read of size 1
==13592==    at 0x10001C77D: TMS320C64x_post_printer (TMS320C64xInstPrinter.c:73)
==13592==    by 0x100013BF0: fill_insn (cs.c:323)
==13592==    by 0x1000134E8: cs_disasm (cs.c:632)
==13592==    by 0x100000E0F: test (in ./test_tms320c64x)
==13592==    by 0x100000CF3: main (in ./test_tms320c64x)
==13592==  Address 0x104801e4f is on thread 1's stack
==13592==  129 bytes below stack pointer
==13592==
==13592== Invalid read of size 1
==13592==    at 0x10001C7BD: TMS320C64x_post_printer (TMS320C64xInstPrinter.c:75)
==13592==    by 0x100013BF0: fill_insn (cs.c:323)
==13592==    by 0x1000134E8: cs_disasm (cs.c:632)
==13592==    by 0x100000E0F: test (in ./test_tms320c64x)
==13592==    by 0x100000CF3: main (in ./test_tms320c64x)
==13592==  Address 0x104801d7d is on thread 1's stack
==13592==  339 bytes below stack pointer
```

**Comment by radare** [comment]
pls update this pr with those fixes and get it merged into next asap. there are several other PRs that must be merged. and the cs4 release is taking SOOOOO MUCH time, im tired of having to carry my own copy of capstone from git everytime i need to do anything useful.

File: cs.c [hunk 1]
```diff
@@ -1065,6 +1065,11 @@ int cs_op_count(csh ud, const cs_insn *insn, unsigned int op_type)
 				if (insn->detail->xcore.operands[i].type == (xcore_op_type)op_type)
 					count++;
 			break;
+		case CS_ARCH_M68K:
```
Comment (radare) [inline cs.c]: probably use a separate PR for this, unrelated to tms320

**Comment by radare** [comment]
please rebase. and will be great if you could fix any of the vulnerabilities i have reported

**Comment by radare** [comment]
please rebase and fix the vulns i have reported. 1 year old already...

**Comment by aquynh** [comment]
I understand that Fosil is too busy to continue this work. Finally, we are all doing this development for free, with zero benefit. If you care about this arch, you should find someone to fix these issues.

**Comment by radare** [comment]
@fotisl != Fosil

maybe he is busy to update but he could at least answer. Yes i do all this for free too, but this doesn't impides me to fuzz, review code, report stuff and fix things. Yes i do care about this arch, as well as all the others mainly because i think it may help other people, and that's all the reward i expect from free software. I'll put all those @fotisl PRs in my TODO list, shoudlnt take more than 1h to rebase and fix all the issues

**Comment by aquynh** [comment]
Sorry for the spelling, as i answered on my phone.

**Comment by radare** [comment]
It takes more time to find someone to fix issues when you are not the author of something, I expect people to be responsible of what they do and publish, that's why i review code, to let them improve and merge things that are safe and learn in the process. Having contributions die in time just because of lack of interest of the maintainer and lack of time of the guy who send it just feels sad.

**Comment by fotisl** [comment]
@radare Please understand that I have already contacted @aquynh several times about this, but I simply have no time right now to start looking at the issues, and merge any new changes since I don't keep track of the changes at capstone. You may make money out of radare2, but I do this in my minimal free time because I have a job at something completely different. So if you want to profit out of it, saying that radare2 supports another architecture, feel free to spend one hour (you mentioned it would take you this long) working on it.

**Comment by radare** [comment]
i’ll release r2-1.4 on monday, but i’ll focus on that right after release, so hopefully enabling this for r2-1.5 and start the whole testing machinery.

Thanks for replying, i have no personal interest on this architecture, but i see no point in seeing people complaining about the lack of some archs and dead pullreqs floating around. the main reason why i started radare was because nobody was maintaining their PoC tools and it was quite a hell to find decent tools that properly cover different use cases.

**Comment by fotisl** [comment]
@radare For the three bugs you reported.
1. This is not a bug in my code, decodeToMCInst_4 is auto-generated by tablegen.
2. Can you give me the hex you tried to decode?
3. Are you sure this is the correct hex?

**Comment by radare** [comment]
i should run the fuzzer again, i dont have the reproducer handy, but it took less than a minute to reproduce. so it shouldnt be hard. i think someone posted a fuzzer for capstone just testing all possible 32bit combinations. that may work, im currently busy with other things. awesome to see you back :)

**Comment by fotisl** [comment]
@radare Any pointers on the fuzzer? I can add an extra check at the place where the third crash takes place, but I don't think it's needed, at least for valid assembly code.

**Comment by radare** [comment]
see suite/fuzz.py

that should work. what i did was in C just modifying the cstool with a for loop incrementing an uint32 until -1

**Comment by aquynh** [comment]
could you please send all the new fixes to the branch https://github.com/aquynh/capstone/tree/tms320c64x?

**Comment by radare** [comment]
Sure, will do after release. Probably on monday. Thanks for the updates

**Comment by fotisl** [comment]
@aquynh I have rebased and corrected the bug (doesn't crash - checked more than 70% of all possible hex values). Python bindings work too. I think it can be merged.

**Comment by aquynh** [comment]
i just merged & pushed to the new branch "tms320c64x-2". @radare, can you do some more tests on that?

**Comment by fotisl** [comment]
Btw, my test completed, I made a loop as @radare suggested and disassembled every possible uint32_t. No problems at all.

**Comment by radare** [comment]
❤️

**Comment by radare** [comment]
Couldnt resist to try it out. The code looks good and could reproduce any crash so far.

The only missing thing i see here is the lack of cstool support, which should be easy to add.

And the only complain I have here is that the disassembler spits tabs, but that's an issue that should be addressed in all disassemblers, i'll do a pullreq and an issue to track this later. But if we get the cstool support looks good to be merged for me.

Good work!

**Comment by aquynh** [comment]
yes, could you please add support for this arch to cstool, then i will merge to the next branch?

**Comment by aquynh** [comment]
merged, and thanks for the excellent work, @fotisl!

**Pull request merged**